cmd/kube-proxy/app/server.go

```diff
@@ -1017,14 +1026,14 @@ func detectNodeIPs(rawNodeIPs []net.IP, bindAddress string) (v1.IPFamily, map[v1
 	}
 
 	if nodeIPs[primaryFamily].IsLoopback() {
-		klog.InfoS("Can't determine this node's IP, assuming loopback; if this is incorrect, please set the --bind-address flag")
+		logger.Info("Can't determine this node's IP, assuming loopback; if this is incorrect, please set the --bind-address flag")
 	}
 	return primaryFamily, nodeIPs
 }
 
 // getNodeIP returns IPs for the node with the provided name.  If
 // required, it will wait for the node to be created.
-func getNodeIPs(client clientset.Interface, name string) []net.IP {
+func getNodeIPs(logger klog.Logger, client clientset.Interface, name string) []net.IP {
```

Comment:
It feels wrong to make this function have to take a logger as an argument...

Basically, this function was previously taking advantage of the existence of a global logger that it could just use without consequences, but if you're going to actually force that logger to be part of its API, then maybe we want to revisit that (and remove the log messages, or else have it return a proper `error` or something else). Failing that, it seems like it would be better to have this function (and similar ones) just grab the global/default logger and have a FIXME comment... (Or alternatively, maybe make them methods of ProxyServer or something?)

I imagine this would come up in other components too? Though I guess kube-proxy is one of the first components to be migrated to the new APIs, right, so we're just discovering this stuff?

Reply:
> Failing that, it seems like it would be better to have this function (and similar ones) just grab the global/default logger and have a FIXME comment...

ah, skimming https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md it seems like maybe this is a job for `klog.TODO()`?

Reply:
Thanks! 
I suggest two ways to solve this problem.
1. Change the function into method of `ProxyServer`, since these functions `getNodeIPs`, `detectNodeIPs` and others are only called in `cmd/kube-proxy/app` package, and all have several INFO logs needed to be output inside them.  And basically, they are all used to initialize some field of `ProxyServer` with other existing fields of `ProxyServer`
2. Change the way of getting logger into `klog.TODO`

WDYT? @danwinship @pohly

Reply:
If we use `klog.TODO()`, what is the final solution? We just delay the decision :hot_face: 

We still have lots of `context.TODO` all over the place, years after client-go started to require a context. Let's not add more technical debt.

> I imagine this would come up in other components too? Though I guess kube-proxy is one of the first components to be migrated to the new APIs, right, so we're just discovering this stuff?

This is not at all new. We had various cases like this in other components. It's always a decision between "add logger" / "add context" / "add logger to instance" - but that last option is often not good because the caller might have a logger with additional key/value pairs that then get ignored when calling a method which uses some other logger that was stored previously.

In a few cases we have decided that logging occurs so infrequently (like dumping a  message in a code path which should never get reached) that changing APIs is not worth it. Then `klog.Background()` is fine, with a comment explaining this.

Reply:
Looking at the code, I'm leaning towards adding `klog.Logger` as parameter.

The alternative is to not log in `getNodeIPs`, but I am not sure whether these messages are really not important to users.

Reply:
> If we use `klog.TODO()`, what is the final solution? We just delay the decision 🥵

Yeah, I was just thinking that this is a question about how to refactor SIG Network's code, so it may make more sense for SIG Instrumentation to just point out that it needs refactoring, rather than trying to guess what we want.

But I get the point about `context.TODO()`

Reply:
> Looking at the code, I'm leaning towards adding `klog.Logger` as parameter.

That's fine for now. I might rewrite it later. :slightly_smiling_face:

Reply:
> it may make more sense for SIG Instrumentation to just point out that it needs refactoring

That's of course would be ideal. It sounds like passing a logger is okay here, but if we encounter code where it's clearly the wrong thing to do, then a `klog.TODO` with clear comments about what to do next is also fine.